src/vm/castcache.cpp

```diff
@@ -0,0 +1,212 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+//
+// File: castcache.h
+//
+
+#include "common.h"
+#include "castcache.h"
+
+CastCache* CastCache::s_cache = NULL;
+
+// this is only called on GC or during shutdown (see: SyncClean::CleanUp).
+CastCache::~CastCache()
+{
+    LIMITED_METHOD_CONTRACT;
+    delete[] m_Table;
+}
+
+CastCache* CastCache::MaybeReplaceCacheWithLarger(CastCache* currentCache)
+{
+    WRAPPER_NO_CONTRACT;
+
+    if (currentCache == NULL)
+    {
+        CastCache* newCache = new (nothrow) CastCache(INITIAL_CACHE_SIZE);
+        if (newCache != NULL)
+        {
+            s_cache = newCache;
+        }
+
+        return newCache;
+    }
+
+    CastCache* newCache = new (nothrow) CastCache(currentCache->Size() * 2);
+    if (newCache == NULL)
+        return FALSE; // oh well, continue using the old cache.
+
+    CastCache* prevCache = FastInterlockCompareExchangePointer(&s_cache, newCache, currentCache);
+
+    if (prevCache == currentCache)
+    {
+        SyncClean::AddObsoleteCastCache(prevCache);
+        return newCache;
+    }
+    else
+    {
+        SyncClean::AddObsoleteCastCache(newCache);
+        return prevCache;
+    }
+}
+
+void CastCache::FlushCurrentCache()
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_PREEMPTIVE;
+    }
+    CONTRACTL_END;
+
+    CastCache* oldCache = s_cache;
+    int newSize = oldCache == NULL ? INITIAL_CACHE_SIZE : oldCache->Size();
+
+    CastCache* newCache = new (nothrow) CastCache(newSize);
+    if (newCache == NULL)
+        newCache = new (nothrow) CastCache(INITIAL_CACHE_SIZE);
+
+    CastCache* obsoleteCache = FastInterlockExchangePointer(&s_cache, newCache);
+
+    if (obsoleteCache != NULL)
+    {
+        SyncClean::AddObsoleteCastCache(obsoleteCache);
+    }
+}
+
+TypeHandle::CastResult CastCache::TryGet(TADDR source, TADDR target)
+{
+    WRAPPER_NO_CONTRACT;
```

Comment:
This needs to have proper contracts (e.g this one needs to be `MODE_COOPERATIVE`) since you depend on GC to be doing synchronization for you.

Reply:
I could not find a good document on contracts, so I borrowed from similar looking methods. I would appreciate some help here.

`Get` does not throw or trigger a gc or care if gc is happening.

I am not sure how that should look in contracts. I saw several different patterns in existing code, but not sure if there are preferred ways.

Reply:
> I could not find a good document on contracts

https://github.com/dotnet/coreclr/blob/master/Documentation/coding-guidelines/clr-code-guide.md#2.10

> `Get` does not throw or trigger a gc or care if gc is happening.

It does care - because of you depend on synchronization with GC to free the unmanaged array.

Reply:
Thanks!  
I thought BoTR + comments in code is all that we have on this matter. Very good document.

Yes it should be `MODE_COOPERATIVE`